File: src/trees/LevelOrderTraversal.java
package com.rohit.trees;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class LevelOrderTraversal {

	public List<List<Integer>> levelOrder(TreeNode root) {

		List<List<Integer>> result = new ArrayList<List<Integer>>();
		Queue<TreeNode> mQueue = new LinkedList<TreeNode>();
		mQueue.add(root);

		while (!mQueue.isEmpty()) {

			int i = 0;
			int size = mQueue.size();
			List<Integer> tempList = new ArrayList<>();
			for (TreeNode tempNode : mQueue) {
				tempList.add(tempNode.val);
			}
			result.add(tempList);
			while (i < size) {

				TreeNode node = mQueue.poll();
				if (node.left != null) {
					mQueue.add(node.left);
				}
				if (node.right != null) {
					mQueue.add(node.right);
				}
				i++;
			}

		}
		return result;
	
	}

}

class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;

	TreeNode(int x) {
		val = x;
	}
}

File: src/trees/LevelOrderTraversal.kt
package src.trees

import java.util.*
import kotlin.collections.ArrayList


fun main() {
    val treeNode = TreeNode(2)
    levelOrder(treeNode)
}


fun levelOrder(root:TreeNode?):List<List<Int>> {
    var result = mutableListOf<List<Int>>()
    var queue = LinkedList<TreeNode?>()
    if (root == null)
         return emptyList()
    queue.add(root)
    queue.add(null)
    var list = ArrayList<Int>()
    while (!queue.isEmpty())
    {
        val node = queue.poll()
        if (node == null)
        {
            result.add(list)
            list = ArrayList<Int>()
            if (queue.isEmpty() || (queue.size == 1 && queue.peek() == null))
                break
            queue.add(null)
            continue
        }
        else
        {
            list.add(node.`val`)
        }
        if (node.left != null)
            queue.add(node.left)
        if (node.right != null)
            queue.add(node.right)
    }
    return result
}